Fix B and C linting rules

The current linter does not seem to catch "B" and "C" rules even though they are enabled in `.flake8`. Running the ruff linter caught the following:

```
movement/laplacian.py:26:45: B006 Do not use mutable data structures for argument defaults
movement/laplacian.py:41:65: B006 Do not use mutable data structures for argument defaults
movement/monge_ampere.py:216:17: B028 No explicit `stacklevel` keyword argument found
movement/monge_ampere.py:547:13: B904 Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling
movement/spring.py:206:65: B006 Do not use mutable data structures for argument defaults
movement/tangling.py:67:13: B028 No explicit `stacklevel` keyword argument found
test/test_smoothing.py:50:9: B007 Loop control variable `j` not used within loop body
Found 7 errors.
```

Now they are
```
movement/math.py:37:24: B023 Function definition does not bind loop variable `hyperplane`
movement/monge_ampere.py:89:9: B904 Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling
movement/monge_ampere.py:157:13: B028 No explicit `stacklevel` keyword argument found
movement/mover.py:47:13: B028 No explicit `stacklevel` keyword argument found
movement/mover.py:68:13: B028 No explicit `stacklevel` keyword argument found
movement/spring.py:180:9: C901 `assemble_stiffness_matrix` is too complex (12 > 10)
movement/tangling.py:67:13: B028 No explicit `stacklevel` keyword argument found
test/test_forced_movement.py:24:20: B006 Do not use mutable data structures for argument defaults
```

--- a/movement/math.py
+++ b/movement/math.py
@@ -33,7 +33,7 @@ def equation_of_hyperplane(*points):
         try:
             hyperplane = Hyperplane(*(Point(points[i]) for i in indices[:dim]))
 
-            def equation(*xyz):
+            def equation(*xyz, hyperplane=hyperplane):
                 return hyperplane.distance(Point(xyz))
 
             return equation

--- a/movement/monge_ampere.py
+++ b/movement/monge_ampere.py
@@ -157,7 +157,7 @@ class MongeAmpereMover_Base(PrimeMover, metaclass=abc.ABCMeta):
             warn(
                 "Provided mesh has no boundary segments with Physical ID tags. If the "
                 "boundaries aren't fully periodic then this will likely cause errors.",
-                stacklevel=4,
+                stacklevel=1,
             )
         elif (
             len(self.fixed_boundary_segments) == 1

--- a/movement/monge_ampere.py
+++ b/movement/monge_ampere.py
@@ -85,8 +85,8 @@ def MongeAmpereMover(mesh, monitor_function, method="relaxation", **kwargs):
     }
     try:
         return implemented_methods[method](mesh, monitor_function, **kwargs)
-    except KeyError:
-        raise ValueError(f"Method '{method}' not recognised.")
+    except KeyError as e:
+        raise ValueError(f"Method '{method}' not recognised.") from e
 
 
 def tangential(v, n):
@@ -156,7 +156,8 @@ class MongeAmpereMover_Base(PrimeMover, metaclass=abc.ABCMeta):
         if len(self._all_boundary_segments) == 0:
             warn(
                 "Provided mesh has no boundary segments with Physical ID tags. If the "
-                "boundaries aren't fully periodic then this will likely cause errors."
+                "boundaries aren't fully periodic then this will likely cause errors.",
+                stacklevel=4,
             )
         elif (
             len(self.fixed_boundary_segments) == 1

--- a/movement/mover.py
+++ b/movement/mover.py
@@ -46,7 +46,8 @@ class PrimeMover(abc.ABC):
         if not raise_convergence_errors:
             warn(
                 f"{type(self)}.move called with raise_convergence_errors=False."
-                " Beware: this option can produce poor quality meshes!"
+                " Beware: this option can produce poor quality meshes!",
+                stacklevel=5,
             )
         self.raise_convergence_errors = raise_convergence_errors
         self.dim = self.mesh.topological_dimension()
@@ -65,7 +66,7 @@ class PrimeMover(abc.ABC):
             self._local_coordinates_vec = dm_coords.createLocalVec()
             self._update_plex_coordinates()
         except ValueError:
-            warn("Cannot update DMPlex coordinates for periodic meshes.")
+            warn("Cannot update DMPlex coordinates for periodic meshes.", stacklevel=5)
             self._local_coordinates_vec = None
 
         self.dx = firedrake.dx(domain=self.mesh, degree=quadrature_degree)

--- a/movement/mover.py
+++ b/movement/mover.py
@@ -47,7 +47,7 @@ class PrimeMover(abc.ABC):
             warn(
                 f"{type(self)}.move called with raise_convergence_errors=False."
                 " Beware: this option can produce poor quality meshes!",
-                stacklevel=5,
+                stacklevel=1,
             )
         self.raise_convergence_errors = raise_convergence_errors
         self.dim = self.mesh.topological_dimension()
@@ -66,7 +66,7 @@ class PrimeMover(abc.ABC):
             self._local_coordinates_vec = dm_coords.createLocalVec()
             self._update_plex_coordinates()
         except ValueError:
-            warn("Cannot update DMPlex coordinates for periodic meshes.", stacklevel=5)
+            warn("Cannot update DMPlex coordinates for periodic meshes.", stacklevel=1)
             self._local_coordinates_vec = None
 
         self.dx = firedrake.dx(domain=self.mesh, degree=quadrature_degree)

--- a/movement/spring.py
+++ b/movement/spring.py
@@ -186,10 +186,9 @@ class SpringMover_Base(PrimeMover):
         :returns: the stiffness matrix with boundary conditions applied
         :rtype: :class:`numpy.ndarray`
         """
-        if not boundary_conditions:
-            boundary_conditions = firedrake.DirichletBC(
-                self.coord_space, 0, "on_boundary"
-            )
+        boundary_conditions = boundary_conditions or firedrake.DirichletBC(
+            self.coord_space, 0, "on_boundary"
+        )
         if isinstance(boundary_conditions, firedrake.DirichletBC):
             boundary_conditions = [boundary_conditions]
         assert isinstance(boundary_conditions, Iterable)
@@ -208,8 +207,7 @@ class SpringMover_Base(PrimeMover):
             tags = boundary_condition.sub_domain
             if tags == "on_boundary":
                 tags = bnd.unique_markers
-            if not isinstance(tags, Iterable):
-                tags = [tags]
+            tags = [tags] if not isinstance(tags, Iterable) else tags
             if not set(tags).issubset(set(bnd.unique_markers)):
                 raise ValueError(f"{tags} contains invalid boundary tags.")
             subsets = np.array([bnd.subset(tag).indices for tag in tags]).flatten()

--- a/movement/tangling.py
+++ b/movement/tangling.py
@@ -64,5 +64,5 @@ class MeshTanglingChecker:
             msg = f"Mesh has {num_tangled} tangled element{plural}."
             if self.raise_error:
                 raise ValueError(msg)
-            warnings.warn(msg, stacklevel=8)
+            warnings.warn(msg, stacklevel=1)
         return num_tangled

--- a/movement/tangling.py
+++ b/movement/tangling.py
@@ -64,5 +64,5 @@ class MeshTanglingChecker:
             msg = f"Mesh has {num_tangled} tangled element{plural}."
             if self.raise_error:
                 raise ValueError(msg)
-            warnings.warn(msg)
+            warnings.warn(msg, stacklevel=8)
         return num_tangled

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -38,8 +38,8 @@ line-length = 88
 
 [tool.ruff.lint]
 select = [
-#  "B",  # flake8-bugbear  TODO: enable this  (#72)
-#  "C",  # mccabe complexity  TODO: enable this (#72)
+  "B",  # flake8-bugbear
+  "C",  # mccabe complexity
   "E", "W",  # Pycodestyle
   "F",  # Pyflakes
   "I",  # isort

--- a/test/monitors.py
+++ b/test/monitors.py
@@ -10,6 +10,6 @@ def ring_monitor(mesh):
     beta = Constant(200.0)  # width
     gamma = Constant(0.15)  # radius
     dim = mesh.geometric_dimension()
-    xyz = SpatialCoordinate(mesh) - as_vector([0.5]*dim)
+    xyz = SpatialCoordinate(mesh) - as_vector([0.5] * dim)
     r = dot(xyz, xyz)
     return Constant(1.0) + alpha / cosh(beta * (r - gamma)) ** 2

--- a/test/test_forced_movement.py
+++ b/test/test_forced_movement.py
@@ -21,9 +21,10 @@ class BaseClasses:
             mesh,
             fixed_boundary_tags=None,
             moving_boundary_tags=None,
-            vector=[1, 0],
+            vector=None,
             **kwargs,
         ):
+            vector = vector or [1, 0]
             mover = self.mover(mesh)
             bcs = []
             if fixed_boundary_tags:

